src/detours.cpp

```diff
@@ -929,6 +930,79 @@ inline void write_opcode(PBYTE &pbCode, ULONG Opcode)
     pbCode += 4;
 }
 
+struct ARM64_INDIRECT_JMP {
+    struct {
+        ULONG Rd : 5;
+        ULONG immhi : 19;
+        ULONG iop : 5;
+        ULONG immlo : 2;
+        ULONG op : 1;
+    } ardp;
+
+    struct {
+        ULONG Rt : 5;
+        ULONG Rn : 5;
+        ULONG imm : 12;
+        ULONG opc : 2;
+        ULONG iop1 : 2;
+        ULONG V : 1;
+        ULONG iop2 : 3;
+        ULONG size : 2;
+    } ldr;
+
+    ULONG br;
+};
+
+#pragma warning(push)
+#pragma warning(disable:4201)
+
+union ARM64_INDIRECT_IMM {
+    struct {
+        ULONG64 pad : 12;
+        ULONG64 adrp_immlo : 2;
+        ULONG64 adrp_immhi : 19;
+    };
+
+    LONG64 value;
+};
+
+#pragma warning(pop)
+
+PBYTE detour_gen_jmp_indirect(BYTE *pbCode, ULONG64 *pbJmpVal)
+{
+    // adrp x17, [jmpval]
```

Comment:
This doesn't quite match the ARM assembler syntax.  It should be // adrp x17, jmpval,  where jmpval is interpreted as a label, i.e. an immediate.    I thinhk it'd be helpful to add a comment to the effect calculating the address of jmpval is split across two instructions.  The first one adrp (loads most significant bits) the second ldr (load the least significant bits of jmpval).

Reply:
Sure, we can add a comment here about the split loading. However this is the canonical and standard way of loading quantities bigger than a page. That's just how things are always done on ARM64. So is it worth it?

Reply:
This is just, "fixed size 32bit instructions require multiple instructions to load a 32bit or 64bit immediate"? Yeah, seems obvious. Maybe some day we will forget x86 and doing it in a single instruction will seem weird. :)